safedelete/queryset.py

```diff
@@ -0,0 +1,133 @@
+
+        A model's m2m field does not easily have access to `all_objects` and
+        so setting `force_visibility` to True is a way of getting all of the
+        models. It is not recommended to use `force_visibility` outside of related
+        models because it will create a new queryset.
+
+        Args:
+            force_visibility: Force a deletion visibility. (default: {None})
+        """
+        if force_visibility is not None:
+            self._safedelete_force_visibility = force_visibility
+        return super(SafeDeleteQueryset, self).all()
+
+    def _check_field_filter(self, **kwargs):
+        """Check if the visibility for DELETED_VISIBLE_BY_FIELD needs t be put into effect.
+
+        DELETED_VISIBLE_BY_FIELD is a temporary visibility flag that changes
+        to DELETED_VISIBLE once asked for the named parameter defined in
+        `_safedelete_force_visibility`. When evaluating the queryset, it will
+        then filter on all models.
+        """
+        if self._safedelete_visibility == DELETED_VISIBLE_BY_FIELD \
+                and self._safedelete_visibility_field in kwargs:
+            self._safedelete_force_visibility = DELETED_VISIBLE
+
+    def filter(self, *args, **kwargs):
+        self._check_field_filter(**kwargs)
+        return super(SafeDeleteQueryset, self).filter(*args, **kwargs)
+
+    def get(self, *args, **kwargs):
+        self._check_field_filter(**kwargs)
+        return super(SafeDeleteQueryset, self).get(*args, **kwargs)
+
+    def _filter_visibility(self):
+        """Add deleted filters to the current QuerySet.
+
+        Unlike QuerySet.filter, this does not return a clone.
+        This is because QuerySet._fetch_all cannot work with a clone.
+        """
+        force_visibility = getattr(self, '_safedelete_force_visibility', None)
+        visibility = force_visibility \
+            if force_visibility is not None \
+            else self._safedelete_visibility
+        if visibility in (DELETED_INVISIBLE, DELETED_VISIBLE_BY_FIELD, DELETED_ONLY_VISIBLE):
+            assert self.query.can_filter(), \
+                "Cannot filter a query once a slice has been taken."
+
+            # Add a query manually, QuerySet.filter returns a clone.
+            # QuerySet._fetch_all cannot work with clones.
+            self.query.add_q(
+                Q(
+                    deleted__isnull=visibility in (
+                        DELETED_INVISIBLE, DELETED_VISIBLE_BY_FIELD
+                    )
+                )
+            )
+
+    def __getattribute__(self, name):
+        """Methods that do not return a QuerySet should call ``_filter_visibility`` first."""
+        attr = object.__getattribute__(self, name)
+        # These methods evaluate the queryset and therefore need to filter the
+        # visiblity set.
+        evaluation_methods = (
+            '_fetch_all', 'count', 'exists', 'aggregate', 'update', '_update',
+            'delete', 'undelete',
```

Comment:
Wow 🎉 
Looked into the code of `django.models.query` and saw it hit `_fetch_all`, upon looking at the method it looks like this: ```    def _fetch_all(self):
        if self._result_cache is None:
            self._result_cache = list(self.iterator())
        if self._prefetch_related_lookups and not self._prefetch_done:
            self._prefetch_related_objects()```

So, fetching all if cached just by the pervious iterator, otherwise prefetch related objects. Prefetching usually means that it's a sliced queryset (correct me if I'm wrong). Changing the `evaluation_methods` `_fetch_all` to `_prefetch_related_objects` worked for me.